frontend/src/query_builder/ExpressionWidget.jsx

```diff
@@ -0,0 +1,465 @@
+        }
+    }
+}
+
+// return the first token with a non-empty error message
+function getErrorToken(tokens) {
+    for (var i = 0; i < tokens.length; i++) {
+        let token = tokens[i];
+        if (token.error && token.error.length) return token;
+        if (!token.isParent) continue;
+        let childError = getErrorToken(token.value);
+        if (childError) return childError;
+    }
+}
+
+function getParsedExpression(tokens) {
+    console.log('getParsedExpression(', tokens, ')');
+
+    if (!tokens || tokens.constructor !== Array || tokens.length !== 3) return null;
+
+    var [lhs, operator, rhs] = tokens;
+
+    if (lhs.error)      throw lhs.error;
+    if (operator.error) throw operator.error;
+    if (rhs.error)      throw rhs.error;
+
+     operator = operator.parsedValue;
+     lhs = lhs.isParent ? getParsedExpression(lhs.value) : lhs.parsedValue;
+     rhs = rhs.isParent ? getParsedExpression(rhs.value) : rhs.parsedValue;
+
+    if (!operator) throw 'invalid operator!';
+    if (!lhs)      throw 'invalid lhs!';
+    if (!rhs)      throw 'invalid rhs!';
+
+    return [operator, lhs, rhs];
+}
+
+
+export default class ExpressionWidget extends Component {
+    constructor(props, context) {
+        super(props, context);
+        _.bindAll(this, 'parseToken', 'parseExpression', 'updateName', 'onExpressionInputChange', 'removeExpression', 'getSuggestions', 'onExpressionInputKeyDown',
+                        'onExpressionInputBlur');
+    }
+
+    static propTypes = {
+        tableMetadata:    PropTypes.object.isRequired,
+        updateExpression: PropTypes.func.isRequired,
+        updateName:       PropTypes.func.isRequired,
+        removeExpression: PropTypes.func.isRequired,
+        name:             PropTypes.string.isRequired,
+        expression:       PropTypes.array.isRequired    // should be an array like [parsedExpressionObj, expressionStringi
+    };
+
+    componentWillMount() {
+        this.componentWillReceiveProps(this.props);
+    }
+
+    componentWillReceiveProps(newProps) {
+        let parsedExpression = newProps.expression[0],
+            expression       = newProps.expression[1],
+            tokens           = expression && expression.length ? getParsedExpression(this.parseExpression(tokenizeExpression(expression))) : []
+        this.setState({
+            name:                   newProps.name,
+            parsedExpression:       parsedExpression,
+            expressionString:       expression,
+            tokens:                 tokens,
+            nameErrorMessage:       newProps.name.length ? null : ERROR_MESSAGE_EMPTY_NAME,
+            expressionErrorMessage: expression.length    ? null : ERROR_MESSAGE_EMPTY_EXPRESSION,
+            suggestions:            [],
+            highlightedSuggestion:  0,
+            suggestionsTitle:       null
+        });
+
+        console.log('component recieved props, state is now: ', this.state);
+    }
+
+    componentWillUnmount() {
```

Comment:
I'm not sure I understand the need for this code.  Why are we doing anything on componentWillUnmount?